components/k8s-model-server/istio-integration.md

```diff
@@ -0,0 +1,42 @@
+# Istio integration for TF serving
+
+[Istio](https://istio.io/) provides a lot of functionalities that we want to have, such as metrics, auth and
+quota, rollout and A/B testing. We have an [issue](https://github.com/kubeflow/kubeflow/issues/464) to track
+the progress.
+
+## Install Istio
+Following the istio [doc](https://istio.io/docs/setup/kubernetes/quick-start.html#installation-steps).
+After the installation, you should see services istio-pilot and istio-mixer in namespace istio-system.
+
+## Kubeflow TF Serving with Istio
+
+This document is working in progress.
+
+Istio by default [denies egress traffic](https://istio.io/docs/tasks/traffic-management/egress.html).
+Since TF serving component might need to read model files from outside (GCS, S3 etc), we need some
+cloud-specific [setting](https://istio.io/docs/tasks/traffic-management/egress.html#calling-external-services-directly). 
+Currently it's for GCP only.
+
+After installing Istio, we can deploy the TF Serving component as in [README](README.md) with
+additional params:
+```
+ks param set --env=cloud ${MODEL_COMPONENT} cloud gcp
```

Comment:
Why not just use ISTIO's admission controller to do automatic sidecar injection rather than doing it manually via config?

https://istio.io/docs/setup/kubernetes/sidecar-injection.html#automatic-sidecar-injection

Reply:
This is an option. This will inject to all pods. Do other kubeflow components want this injection?

Reply:
It can be set as a param of init container, so should be per pod.

Reply:
Per our offline discussion I think we should try to use the admission controller if at all possible. The ideal would be to see if we can target the admission controller to match pods we want to include vs. exclude.

I think admission controllers are a better pattern for doing these types of modifications. In part because I think the admission controller for ISTIO is probably not something we want to manage as part of Kubeflow because it will probably need to be customized in very specific ways.

The ip range for egress routes/rules is one example. 

IIUC one of the goals for ISTIO is to allow policies to be centrally maintained so that a cluster admin can enforce certain constraints like ingress/egress filters.